sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala

```diff
@@ -130,8 +130,7 @@ private[hive] object IsolatedClientLoader extends Logging {
     }
     val hiveArtifacts = version.extraDeps ++
       Seq("hive-metastore", "hive-exec", "hive-common", "hive-serde")
-        .map(a => s"org.apache.hive:$a:${version.fullVersion}") ++
-      Seq("com.google.guava:guava:14.0.1") ++ hadoopJarNames
+        .map(a => s"org.apache.hive:$a:${version.fullVersion}") ++ hadoopJarNames
```

Comment:
Here, I think the Guava version needs to be a function of `version`, rather than matching Spark's Guava version: if we remove Guava from shared classes but also changed the now-non-shared metastore Guava version then it's the same net effect and breaks older Hive versions.

I believe that newer versions of Hive shade Guava (in which case they're insensitive to whatever value we set here). I think we could either (a) continue to unconditionally use Guava 14.0.1 here, or (b) conditionally use it only for older Hive versions that predated the Guava shading.

Reply:
> if we remove Guava from shared classes but also changed the now-non-shared metastore Guava version then it's the same net effect and breaks older Hive versions.

Hmm, I don't get the idea. After we remove Guava from the shared class, then we don't need to keep Hive and Spark using the same Guava. And we did not add Guava to `HiveVersion#exclusions`, each Hive version should automatically pull all transitive dependencies including Guava if necessary.

Anyway, let's simply try (a) first.

Reply:
@JoshRosen Unfortunately, it does not work.
```
Caused by: sbt.ForkMain$ForkError: java.lang.AssertionError: java.lang.NoSuchMethodException: com.google.common.base.internal.Finalizer.startFinalizer(java.lang.Class, java.lang.Object)
	at com.google.common.base.FinalizableReferenceQueue.getStartFinalizer(FinalizableReferenceQueue.java:283)
	at com.google.common.base.FinalizableReferenceQueue.<clinit>(FinalizableReferenceQueue.java:86)
	at com.jolbox.bonecp.BoneCP.<init>(BoneCP.java:427)
	at com.jolbox.bonecp.BoneCPDataSource.getConnection(BoneCPDataSource.java:120)
	at org.datanucleus.store.rdbms.ConnectionFactoryImpl$ManagedConnectionImpl.getConnection(ConnectionFactoryImpl.java:501)
	at org.datanucleus.store.rdbms.RDBMSStoreManager.<init>(RDBMSStoreManager.java:298)
        ...
```

Reply:
Also, cc @sunchao, I think the experiment indicates that we can not upgrade Guava while keeping `IsolatedClientLoader`

Reply:
things changed after we removed support for Hive prior 2.0.0, we can remove Guava from the shared class list now, thus we can do Guava upgrading while keeping `IsolatedClientLoader `